plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+                pubKeyPtr += sizeof(keySize);
+                publicKeys.push_back(pkeyCache.checkCache(keySize, pubKeyPtr, 0, nullptr));
+                pubKeyPtr += keySize;
+            }
+
+            // Load the cipher
+            const EVP_CIPHER * cipher = cipherCache.checkCache(_algorithm_name);
+
+            // Allocate memory for encrypted keys
+            size_t keyCount = publicKeys.size();
+            encryptedKeys = static_cast<byte **>(rtlMalloc(sizeof(byte *) * keyCount));
+            for (size_t x = 0; x < keyCount; x++)
+                encryptedKeys[x] = static_cast<byte *>(rtlMalloc(EVP_PKEY_size(publicKeys[x])));
+
+            // Allocate memory for the IV
+            int ivLen = EVP_CIPHER_iv_length(cipher);
+            iv.ensureCapacity(ivLen);
+
+            // Allocate buffer for ciphertext
+            int ciphertextLen = len_plaintext + EVP_CIPHER_block_size(cipher);
+            ciphertext.ensureCapacity(ciphertextLen);
+
+            // Create and initialize the context
+            encryptCtx = EVP_CIPHER_CTX_new();
+            if (!encryptCtx)
+                failOpenSSLError("creating cipher context");
+
+            // Initialize the envelope
+            std::vector<int> keyLens(keyCount);
+            if (EVP_SealInit(encryptCtx, cipher, encryptedKeys, keyLens.data(), static_cast<byte *>(iv.bufferBase()), publicKeys.data(), keyCount) != static_cast<int>(keyCount))
+                failOpenSSLError("EVP_SealInit");
+
+            // Update the envelope (encrypt the plaintext)
+            int len = 0;
+            if (EVP_SealUpdate(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()), &len, reinterpret_cast<const byte *>(_plaintext), len_plaintext) != 1)
+                failOpenSSLError("EVP_SealUpdate");
+            ciphertextLen = len;
+
+            // Finalize the envelope's ciphertext
+            if (EVP_SealFinal(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()) + len, &len) != 1)
+                failOpenSSLError("EVP_SealFinal");
+            ciphertextLen += len;
+
+            int totalKeyLen = 0;
+            for (int i = 0; i < keyCount; i++)
+                totalKeyLen += keyLens[i];
+
+            // We need to prepend the ciphertext with metadata so the blob can be decrypted;
+            // this is potentially nonstandard
+            MemoryBuffer outBuffer;
+            outBuffer.ensureCapacity(ivLen + (sizeof(size_t)*(keyCount+1)) + totalKeyLen + ciphertextLen);
+            // IV comes first; its length can be derived from the cipher
+            outBuffer.append(ivLen, static_cast<byte *>(iv.bufferBase()));
+            // Number of keys (size_t)
+            outBuffer.append(sizeof(keyCount), reinterpret_cast<byte *>(&keyCount));
+            // Keys; each is (size_t) + <content>
+            for (size_t x = 0; x < keyCount; x++)
+            {
+                size_t keyLen = keyLens[x];
```

Comment:
Should this be size32_t rather than size_t?  All ecl structures are 32 bit lengths.  Another alternative if it is not read from ecl, you could use appendPacked()